pyccel/codegen/printing/fcode.py

```diff
@@ -452,28 +455,151 @@ def _print_Import(self, expr):
+                arg_format, arg = self.get_print_format_and_arg(f)
+                args_format.append(arg_format)
+                args.append(arg)
             elif isinstance(f, InhomogeneousTupleVariable):
-                for i in f:
-                    args.append(self._print(i))
-            elif f.dtype is NativeString() and j != n-1:
-                args.append("{} // ' ' ".format(self._print(f)))
-            else:
+                arg_format, arg = self.get_print_format_and_arg(f)
+                args_format.append(arg_format)
+                args.append(arg)
+            elif f.dtype is NativeString() and f != expr.expr[-1]:
+                args_format.append('A')
                 args.append(self._print(f))
+            elif isinstance(f.rank, int) and f.rank > 0:
+                if args_format:
+                    code += formatted_args_to_print(args_format, args, self._print_LiteralString(LiteralString(sep)))
+                    args_format = []
+                    args = []
+                for_index = Variable(NativeInteger(), name=self.parser.get_new_name('i'))
+                self._additional_declare.append(for_index)
+                max_index = PyccelMinus(f.shape[0], LiteralInteger(1), simplify=True)
+                for_range = PythonRange(max_index)
+                print_body = [FunctionCallArgument(f[for_index])]
+                if f.rank == 1:
+                    print_body.append(space_end)
+
+                for_body = [PythonPrint(print_body)]
+                for_loop = For(for_index, for_range, for_body)
+                for_end = FunctionCallArgument(LiteralString(']' + end if i == len(orig_args) - 1 else ']'),
+                                               keyword='end')
+
+                body = CodeBlock([PythonPrint([FunctionCallArgument(LiteralString('[')), empty_end]),
+                                  for_loop,
+                                  PythonPrint([FunctionCallArgument(f[max_index]), for_end])],
+                                 unravelled=True)
+                code += self._print(body)
+            else:
+                arg_format, arg = self.get_print_format_and_arg(f)
+                args_format.append(arg_format)
+                args.append(arg)
+        if args_format:
+            code += formatted_args_to_print(args_format, args, end)
+        return code
 
-        code = ', '.join(['print *', *args])
-        return self._get_statement(code) + '\n'
+    def get_print_format_and_arg(self,var):
+        type_to_format = {('real'): 'F0.12',
+                          ('complex'): '"(",F0.12," + ",F0.12,")"',
+                          ('integer'): 'I0',
+                          ('logical'): 'A',
+                          ('string'): 'A'}
+        var_type = self._print(var.dtype)
+        if var_type not in type_to_format:
+            if self._print(var.dtype).find("character") != -1:
+                return 'A', self._print(var)
```

Comment:
Why do you recalculate `self._print(var.dtype)`? Can't you use `var_type.find("character")` ?

What are you trying to handle here? A class with character in the name?

Reply:
This is the tricky part i m not sure about.
I looked into the format specifiers for `write` in Fortran, and was only able to match these:
```
 def get_print_format_and_arg(self,var):
        type_to_format = {('real'): 'F0.12',
                          ('complex'): '"(",F0.12," + ",F0.12,")"',
                          ('integer'): 'I0',
                          ('logical'): 'A',
                          ('string'): 'A'}
```
if the dtype is not present in my dictionary i fall back to the default print with `print *, ...` . I am pretty sure i did not cover all cases. I did not find good enough of a documentation on the matter. 
```suggestion
            if self._print(var.dtype).find("character") != -1:
                return 'A', self._print(var)
```
this is to check if the type is a character then i could use  `A` with it as well.

Reply:
I agree that falling back to `print *, ...` is probably the safest

Reply:
> this is to check if the type is a character then i could use A with it as well.

`character` is a fortran type. `var.dtype` returns the python type collected by pyccel. It will always be one of the types defined in [this file](https://github.com/pyccel/pyccel/blob/master/pyccel/ast/datatypes.py) or the name of a class. I don't see anything named `character` here so I don't see the use case for your code